code/game/turfs/simulated/floor_icon.dm

```diff
@@ -132,28 +135,174 @@ var/list/flooring_cache = list()
+							//Found a match on the list
+							is_linked = TRUE
+							break
+				else if(floor_smooth == SMOOTH_BLACKLIST)
+					is_linked = TRUE //Default to true for the blacklist, then make it false if a match comes up
+					for (var/v in flooring_whitelist)
+						if (istype(t.flooring, v))
+							//Found a match on the list
+							is_linked = FALSE
+							break
+	else
+		//If this is a countercheck, we skip all of the above, start off with true, and go straight to the atom lists
+		is_linked = TRUE
+
+
+
+	//Alright now we have a preliminary answer about smoothing, however that answer may change with the following
+	//Atom lists!
+	var/best_priority = -1
+	//A white or blacklist entry will only override smoothing if its priority is higher than this
+	//And then this value becomes its priority
+	if (smooth_movable_atom != SMOOTH_NONE)
+		if (smooth_movable_atom == SMOOTH_WHITELIST || smooth_movable_atom == SMOOTH_GREYLIST)
+			for (var/list/v in movable_atom_whitelist)
+				var/d_type = v[1]
+				var/list/d_vars = v[2]
+				var/d_priority = v[3]
+				//Priority is the quickest thing to check first
+				if (d_priority <= best_priority)
+					continue
+
+				//Ok, now we start testing all the atoms in the target turf
+				for (var/a in T) //No implicit typecasting here, faster
+
+					if (istype(a, d_type))
+						//It's the right type, so we're sure it will have the vars we want.
+
+						var/atom/movable/AM = a
+						//Typecast it to a movable atom
+						//Lets make sure its in the way before we consider it
+						if (!AM.is_between_turfs(origin, T))
+							continue
+
+						//From here on out, we do dangerous stuff that may runtime if the coder screwed up
+
+
+						var/match = TRUE
+						for (var/d_var in d_vars)
+							//For each variable we want to check
+							if (AM.vars[d_var] != d_vars[d_var])
+								//We get a var of the same name from the atom's vars list.
+								//And check if it equals our desired value
+								match = FALSE
+								break //If any var doesn't match the desired value, then this atom is not a match, move on
+
 
-	//If is_hole is true, then it's space or openspace
-	else if(T.is_hole)
-		if(flooring.space_smooth)
-			is_linked = TRUE
+						if (match)
+							//If we've successfully found an atom which matches a list entry
+							best_priority = d_priority //This one is king until a higher priority overrides it
 
-	//If we get here then its a normal floor
-	else if (istype(T, /turf/simulated/floor))
-		var/turf/simulated/floor/t = T
-		if (flooring.floor_smooth || t.flooring.name == flooring.name)
-			is_linked = TRUE
+							//And this is a whitelist, so this match forces is_linked to true
+							is_linked = TRUE
+
+
+		if (smooth_movable_atom == SMOOTH_BLACKLIST || smooth_movable_atom == SMOOTH_GREYLIST)
+			//All of this blacklist code is copypasted from above, with only minor name changes
+			if (smooth_movable_atom == SMOOTH_BLACKLIST || smooth_movable_atom == SMOOTH_GREYLIST)
```

Comment:
That check is redundand.
![default](https://user-images.githubusercontent.com/2458675/48902591-d1ea1580-ee7a-11e8-9f1b-55038f4a6568.PNG)